### PR TITLE
Convert UPGRADE.rst to markdown

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/10125

This PR:

* Converts UPGRADE.rst to markdown and moves the contents into the `docs/` directory.
* Updates the contents of UPGRADE.rst to point to the website instead.
* Updates links around the codebase that point to UPGRADE.rst.

`pandoc` + some manual editing was used to convert from RST to md.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1985,7 +1985,7 @@ No significant changes since 1.19.0rc1.
 Removal warning
 ---------------
 
-As outlined in the [previous release](https://github.com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker images with the `-py3` tag suffix. On top of that, we have also removed the `latest-py3` tag. Please see [the announcement in the upgrade notes for 1.18.0](https://github.com/matrix-org/synapse/blob/develop/UPGRADE.rst#upgrading-to-v1180).
+As outlined in the [previous release](https://github.com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker images with the `-py3` tag suffix. On top of that, we have also removed the `latest-py3` tag. Please see [the announcement in the upgrade notes for 1.18.0](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#upgrading-to-v1180).
```
Comment (richvdh) [inline CHANGES.md:1988]: should this (and others below) be `/upgrading/README.md` ?

File: README.rst [hunk 2]
```diff
@@ -193,11 +193,11 @@ impact to other applications will be minimal.
 Upgrading an existing Synapse
 =============================
 
-The instructions for upgrading synapse are in `UPGRADE.rst`_.
+The instructions for upgrading synapse are in `the upgrade notes`_.
 Please check these instructions as upgrading may require extra steps for some
 versions of synapse.
 
-.. _UPGRADE.rst: UPGRADE.rst
+.. _the upgrade notes: https://matrix-org.github.io/synapse/develop/upgrading
```
Comment (richvdh) [inline README.rst:200]: `upgrading/README.md` ?

File: UPGRADE.rst [hunk 3]
```diff
@@ -1,1324 +1,7 @@
 Upgrading Synapse
 =================
 
-Before upgrading check if any special steps are required to upgrade from the
-version you currently have installed to the current version of Synapse. The extra
-instructions that may be required are listed later in this document.
+This document has moved to the `Synapse documentation website <https://matrix-org.github.io/synapse/develop/upgrading/>`_.
```
Comment (richvdh) [inline UPGRADE.rst:4]: `upgrading/README.md`?

**Comment by richvdh** [comment]
what is the thinking behind calling this `upgrading/README.md` rather than, say, `upgrading.md`?

File: CHANGES.md (hunk 1 above)
Comment (anoadragon453) [reply CHANGES.md:1988]: Oops, halfway through I switched from `/docs/upgrade.md` -> `/docs/upgrading/README.md`.

**Comment by anoadragon453** [comment]
> what is the thinking behind calling this `upgrading/README.md` rather than, say, `upgrading.md`?

Hmm, just so that it was not top-level, but also not buried too deeply. I wonder if we'd want to make an exception for this document specifically as it may be one that people browsing GitHub/a code checkout frequently look for?

File: README.rst (hunk 2 above)
Comment (anoadragon453) [reply README.rst:200]: mdbook will transform `README.md` files to `index.html`, and github will resolve `/synapse/develop/upgrading` to `/synapse/develop/upgrading/index.html` for us.

File: UPGRADE.rst (hunk 3 above)
Comment (anoadragon453) [reply UPGRADE.rst:4]: Same as above.

**Comment by richvdh** [comment]
> Hmm, just so that it was not top-level, but also not buried too deeply.

Mostly I don't see the advantage of replacing a single top-level file with a top-level directory. ymmv though.

File: CHANGES.md [hunk 4]
```diff
@@ -2022,7 +2026,7 @@ Bugfixes
 Updates to the Docker image
 ---------------------------
 
-- We no longer publish Docker images with the `-py3` tag suffix, as [announced in the upgrade notes](https://github.com/matrix-org/synapse/blob/develop/UPGRADE.rst#upgrading-to-v1180). ([\#8056](https://github.com/matrix-org/synapse/issues/8056))
+- We no longer publish Docker images with the `-py3` tag suffix, as [announced in the upgrade notes](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#upgrading-to-v1180). ([\#8056](https://github.com/matrix-org/synapse/issues/8056))
```
Comment (richvdh) [inline CHANGES.md:2029]: this one's still broken

**Comment by anoadragon453** [comment]
Ah right. The idea was to follow the convention laid forth in https://github.com/matrix-org/synapse/tree/develop/docs#adding-to-the-documentation.

**Comment by richvdh** [comment]
> Ah right. The idea was to follow the convention laid forth in https://github.com/matrix-org/synapse/tree/develop/docs#adding-to-the-documentation. However for this specific instance it's a bit murky.

It's not obvious to me that that says "everything must be in its own directory", but if that is what it says, it feels wrong?

**Comment by anoadragon453** [comment]
> > Ah right. The idea was to follow the convention laid forth in https://github.com/matrix-org/synapse/tree/develop/docs#adding-to-the-documentation. However for this specific instance it's a bit murky.
> 
> It's not obvious to me that that says "everything must be in its own directory", but if that is what it says, it feels wrong?

Oh sorry. The takeaway from the link was that files should be placed in directories that match their structure in `SUMMARY.md`:

https://github.com/matrix-org/synapse/blame/develop/docs/SUMMARY.md#L13-L14

Since this is under the `Upgrading` header, it would live in a folder called `upgrading`.

The second part is that the file should be named according to the title of the document, so `upgrade_synapse.md`. But I thought we'd deviate from that bit by calling it README.md, so that GitHub displays it immediately upon entering the `docs/upgrading` folder.

That's how I ended up here, but I could go either way on this file as I recognise it's one we'll want people to easily find in the repo:

1. `docs/upgrade.md`
2. `docs/upgrading/README.md`
3. `docs/upgrading/upgrade_synapse.md`

**Comment by anoadragon453** [comment]
@richvdh apologies for the lack of a decision here. I think I'm just going to go with `docs/upgrade.md` for now and we can move it into a folder later if we end up having more than one document related to upgrading Synapse :slightly_smiling_face:

File: CHANGES.md [hunk 5]
```diff
@@ -1991,7 +1991,11 @@ No significant changes since 1.19.0rc1.
 Removal warning
 ---------------
 
-As outlined in the [previous release](https://github.com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker images with the `-py3` tag suffix. On top of that, we have also removed the `latest-py3` tag. Please see [the announcement in the upgrade notes for 1.18.0](https://github.com/matrix-org/synapse/blob/develop/UPGRADE.rst#upgrading-to-v1180).
+As outlined in the [previous release](https://github.
+com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker
```
Comment (richvdh) [inline CHANGES.md:1995]: this link is broken by the newline

File: CHANGES.md [hunk 6]
```diff
@@ -1991,7 +1991,11 @@ No significant changes since 1.19.0rc1.
 Removal warning
 ---------------
 
-As outlined in the [previous release](https://github.com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker images with the `-py3` tag suffix. On top of that, we have also removed the `latest-py3` tag. Please see [the announcement in the upgrade notes for 1.18.0](https://github.com/matrix-org/synapse/blob/develop/UPGRADE.rst#upgrading-to-v1180).
+As outlined in the [previous release](https://github.
+com/matrix-org/synapse/releases/tag/v1.18.0), we are no longer publishing Docker 
+images with the `-py3` tag suffix. On top of that, we have also removed the 
+`latest-py3` tag. Please see [the announcement in the upgrade notes for 1.18.0]
+(https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#upgrading-to-v1180).
```
Comment (richvdh) [inline CHANGES.md:1998]: so is this one

File: README.rst [hunk 7]
```diff
@@ -186,11 +186,11 @@ impact to other applications will be minimal.
...
```
Comment (richvdh) [inline README.rst:193]:
```suggestion
.. _the upgrade notes: https://matrix-org.github.io/synapse/develop/docs/upgrade.md
```

I think?

File: changelog.d/10062.removal [hunk 8]
```diff
@@ -1 +1 @@
-The current spam checker interface is deprecated in favour of a new generic modules system. See the [upgrade notes](https://github.com/matrix-org/synapse/blob/master/UPGRADE.rst#deprecation-of-the-current-spam-checker-interface) for more information on how to update to the new system.
\ No newline at end of file
+The current spam checker interface is deprecated in favour of a new generic modules system. See the [upgrade notes](https://github.com/matrix-org/synapse/blob/master/docs/upgrade.md#deprecation-of-the-current-spam-checker-interface) for more information on how to update to the new system.
```
Comment (richvdh) [inline changelog.d/10062.removal:1]: This should match CHANGES.md

```suggestion
The current spam checker interface is deprecated in favour of a new generic modules system. See the [upgrade notes](docs/upgrade.md#deprecation-of-the-current-spam-checker-interface) for more information on how to update to the new system.
```

**Review by richvdh — CHANGES_REQUESTED** [review]

File: changelog.d/10062.removal (hunk 8 above)
Comment (anoadragon453) [reply changelog.d/10062.removal:1]: Relative links won't work in the release notes page though I think.

File: README.rst (hunk 7 above)
Comment (anoadragon453) [reply README.rst:193]: mdbook operates from the perspective of the `docs` folder, so I believe if we put the file at `docs/upgrade.md` then the resulting URL will be `https://matrix-org.github.io/synapse/develop/upgrade.html`.

We can lob off the `.html` bit though.

File: changelog.d/10062.removal (hunk 8 above)
Comment (richvdh) [reply changelog.d/10062.removal:1]: true, I guess.

**Review by richvdh — APPROVED** [review]

**Pull request merged**